### PR TITLE
Feature: implement test challenge

### PR DESCRIPTION
**Pull request by lpdat2796**

### CODE REVIEW

File: app/forms/new_transaction_form.rb [hunk 1]
```diff
@@ -0,0 +1,42 @@
+# frozen_string_literal: true
+
+class NewTransactionForm
```
Comment (abdulmajid-mamopay) [inline app/forms/new_transaction_form.rb:3]: Hmmmmm 🤔 , are form objects ideal for APIs? Maybe service objects could have been better.

File: app/forms/new_transaction_form.rb [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+      transaction.save
+    end
+  rescue ActiveRecord::RecordInvalid => e
+    transaction.status = :failed
+    transaction.message = e.message
+    transaction.save
+    errors.add(:base, e.message)
+
+    false
+  end
```
Comment (abdulmajid-mamopay) [inline app/forms/new_transaction_form.rb:41]: Maybe you could have used private methods to simplify this method

File: app/controllers/api/v1/transactions_controller.rb [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    end
+
+    if @new_transaction_form.save
+      render_response_success(
+        200,
+        @transaction,
+        each_serializer: TransactionSerializer
+      )
+    else
+      render json: { status: 401, message: @new_transaction_form.errors.full_messages }
```
Comment (abdulmajid-mamopay) [inline app/controllers/api/v1/transactions_controller.rb:43]: Hmmm 🤔 , I think 401 still means the user is unauthorized. That doesn't seem right does it?

File: app/controllers/api/v1/transactions_controller.rb [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    end
+  end
+
+  private
+
+  def transaction_params
+    params.permit(:receiver, :transaction_type, :amount, :message)
```
Comment (abdulmajid-mamopay) [inline app/controllers/api/v1/transactions_controller.rb:50]: Nice! A possible improvement here is to use a library to really check if the params are there and of the required types.

File: app/controllers/api/v1/transactions_controller.rb [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+  end
+
+  def set_transaction
+    if params[:id]
```
Comment (abdulmajid-mamopay) [inline app/controllers/api/v1/transactions_controller.rb:54]: A service object will have removed some of these business logic away from the controller

**Review by abdulmajid-mamopay — COMMENTED** [review]
Good job! There are possible areas of improvement like params validation, business logic extraction, better error handling.

Congratulations on the job done 🎉

File: app/forms/new_transaction_form.rb (hunk 1 above)
Comment (lpdat2796) [reply app/forms/new_transaction_form.rb:3]: I think when using FormObject, we can easily handle some complicated logic or condition on action **save** or **destroy** record. Ex: save record on table but need to update/validate in others related tables.
But your idea is also good so I will consider it in future :bow:

File: app/forms/new_transaction_form.rb (hunk 2 above)
Comment (lpdat2796) [reply app/forms/new_transaction_form.rb:41]: Yes, I will fix it, thank you for your comment 🙇

File: app/controllers/api/v1/transactions_controller.rb (hunk 3 above)
Comment (lpdat2796) [reply app/controllers/api/v1/transactions_controller.rb:43]: Yes you right, in this case **400** is correct. Thanks for your comment

File: app/controllers/api/v1/transactions_controller.rb (hunk 4 above)
Comment (lpdat2796) [reply app/controllers/api/v1/transactions_controller.rb:50]: You mean gem **dry-schema**? I have researched this gem but i'm not quite understand it throughly, I will research it more and try to use it in the future

File: app/controllers/api/v1/transactions_controller.rb (hunk 5 above)
Comment (lpdat2796) [reply app/controllers/api/v1/transactions_controller.rb:54]: As I mention in the comment above, I will use gem **dry-schema** in the future

**Comment by lpdat2796** [comment]
> Good job! There are possible areas of improvement like params validation, business logic extraction, better error handling.
> Congratulations on the job done 🎉

Hi @abdulmajid-mamopay, thanks a lot for your feedbacks. I learns a lot from them. 🙇

**Pull request closed without merging**